src/worker/helpers.go

```diff
@@ -166,46 +166,192 @@ func initOLDir(olPath string, dockerBaseImage string, newBase bool) (err error)
+
 	if err := p.Signal(syscall.SIGINT); err != nil {
-		return fmt.Errorf("Failed to send SIGINT to PID %d (%s).  May require manual cleanup.\n", pid, err.Error())
+		fmt.Printf("Failed to send SIGINT to PID %d: %s. Manual cleanup may be required.\n", p.Pid, err.Error())
+		return fmt.Errorf("failed to send SIGINT to PID %d: %s", p.Pid, err.Error())
 	}
 
+	// Check the process status every 100 milliseconds for up to 60 seconds
 	for i := 0; i < 600; i++ {
 		err := p.Signal(syscall.Signal(0))
 		if err != nil {
-			fmt.Printf("OL worker process stopped successfully.\n")
+			fmt.Println("Worker process stopped successfully.")
 			return nil
 		}
 		time.Sleep(100 * time.Millisecond)
 	}
 
-	return fmt.Errorf("worker didn't stop after 60s")
+	return fmt.Errorf("worker process did not stop within 60 seconds")
+}
+
+// This function attempts to clean up resources after detecting a dirty shutdown.
+// It cleans up cgroups and mounts associated with the Open Lambda instance at `olPath`.
+// Returns errors encountered during cleanup operations.
+func dirtyCleanup(olPath string) error {
+	// Clean up cgroups associated with sandboxes
+	cgRoot := filepath.Join("/sys", "fs", "cgroup", filepath.Base(olPath)+"-sandboxes")
+	fmt.Printf("Attempting to clean up cgroups at %s\n", cgRoot)
+
+	if files, err := os.ReadDir(cgRoot); err != nil {
+		// Log an error if the cgroup root directory cannot be found.
+		fmt.Printf("Could not find cgroup root: %s\n", err.Error())
+	} else {
+		kill := filepath.Join(cgRoot, "cgroup.kill")
+		if err := os.WriteFile(kill, []byte(fmt.Sprintf("%d", 1)), os.ModeAppend); err != nil {
+			// Log an error if killing processes in the cgroup fails.
+			fmt.Printf("Could not kill processes in cgroup: %s\n", err.Error())
+		}
+		for _, file := range files {
+			if strings.HasPrefix(file.Name(), "cg-") {
+				cg := filepath.Join(cgRoot, file.Name())
+				fmt.Printf("Attempting to remove %s\n", cg)
+				if err := syscall.Rmdir(cg); err != nil {
+					// Return an error if removing a cgroup fails.
+					return fmt.Errorf("could not remove cgroup: %s", err.Error())
+				}
+			}
+		}
+		if err := syscall.Rmdir(cgRoot); err != nil {
+			// Log an error if removing the cgroup root directory fails.
+			fmt.Printf("Could not remove cgroup root: %s\n", err.Error())
+		}
+	}
+
+	// Clean up mounts associated with sandboxes
+	dirName := filepath.Join(olPath, "worker", "root-sandboxes")
+	fmt.Printf("Attempting to clean up mounts at %s\n", dirName)
+
+	if files, err := os.ReadDir(dirName); err != nil {
+		// Return an error if the mount root directory cannot be found.
+		return fmt.Errorf("could not find mount root: %s", err.Error())
+	} else {
+		for _, file := range files {
+			path := filepath.Join(dirName, file.Name())
+			fmt.Printf("Attempting to unmount %s\n", path)
+			if err := syscall.Unmount(path, syscall.MNT_DETACH); err != nil {
+				// Return an error if unmounting fails.
+				return fmt.Errorf("could not unmount: %s", err.Error())
```

Comment:
If we return now, we won't attempt to unmount the other directories (unmounting those might succeed!)

Reply:
My assumption is the user can run simply the cleanup process again after they resolve what causes those cgroups too unable to be unmounted (either though CLI or through the automatic cleanup process), and early return might be helpful for the users to find what went wrong.
I assume we don't want to simply continue when some cleanup processes were not successful.